### PR TITLE
BAL08 - Large Delegation

### PR DESCRIPTION
**Pull request by Hyodar**
# Balancer Large Delegation Bot

## Types of changes

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [ ] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

### CODE REVIEW

File: Balancer-Bots/Large-Delegation/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Balancer Large Delegation Bot
+
+## Description
+
+This bot detects "large" (in terms of an absolute token amount threshold or a total supply percentage threshold) veBAL
+delegations by monitoring delegations through the `DelegateRegistry` contract and checking veBAL balances.
```
Comment (e1Ru1o) [inline Balancer-Bots/Large-Delegation/README.md:6]: I prefer something like:
```suggestion
This bot detects "large" veBAL delegations by monitoring delegations through the `DelegateRegistry` contract and checking veBAL balances.
> "Large" is defined in terms of an absolute token amount threshold or a total supply percentage threshold 
```

File: Balancer-Bots/Large-Delegation/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+- BAL-8-1
+  - Fired when a large (in absolute terms) delegation happens
+  - Severity is always set to "info"
+  - Type is always set to "info"
+  - Metadata:
+    - `delegator`: The delegator address
+    - `delegate`: The delegate address
+    - `amount`: The delegator veBAL balance
+
+- BAL-8-1
```
Comment (e1Ru1o) [inline Balancer-Bots/Large-Delegation/README.md:25]:
```suggestion
- BAL-8-2
```

File: Balancer-Bots/Large-Delegation/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    const logs = await provider.getLogs({
+      address: config.delegateRegistryAddress,
+      topics,
+      fromBlock: blockEvent.blockNumber,
+      toBlock: blockEvent.blockNumber,
+    });
+
+    const parsedLogs = logs
+      .map((log) => delegateRegistryIface.parseLog(log))
+      .filter((log) => log.args.id === balancerId);
```
Comment (e1Ru1o) [inline Balancer-Bots/Large-Delegation/src/agent.ts:31]: I think you can avoid this filter by adding the id in the topics as it is and indexed argument of the event

**Review by e1Ru1o — APPROVED** [review]

File: Balancer-Bots/Large-Delegation/src/agent.spec.ts [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    mockProvider.addLogs([getSetDelegateLog(DELEGATE_REGISTRY_ADDRESS, 1, ADDRESSES[0], "balancer.eth", ADDRESSES[1])]);
+
+    expect(await handleBlock(blockEvent)).toStrictEqual([]);
+    expect(mockProvider.call).not.toHaveBeenCalled();
+  });
+
+  it("should ignore delegations whose id is not 'balancer.eth'", async () => {
+    const blockEvent = new TestBlockEvent().setNumber(0);
+
+    mockProvider.addLogs([getSetDelegateLog(DELEGATE_REGISTRY_ADDRESS, 1, ADDRESSES[0], "some.id", ADDRESSES[1])]);
```
Comment (0xtaf) [inline Balancer-Bots/Large-Delegation/src/agent.spec.ts:150]:
```suggestion
    mockProvider.addLogs([getSetDelegateLog(DELEGATE_REGISTRY_ADDRESS, 0, ADDRESSES[0], "some.id", ADDRESSES[1])]);
```
very small thing. it just makes more sense to give the same block number as it tests the effect of different ids

**Review by 0xtaf — APPROVED** [review]
good job again 💯

File: Balancer-Bots/Large-Delegation/src/agent.spec.ts (hunk 4 above)
Comment (Hyodar) [reply Balancer-Bots/Large-Delegation/src/agent.spec.ts]: Yup, good catch.

**Pull request merged**